### PR TITLE
Allow circular references

### LINKED ISSUE #64: Bug: Circular field dependency results in maximum call stack
This code:
```js
import {
	Schema,
	Query,
	ObjectType,
	Field,
	schemaFactory
} from 'graphql-schema-decorator'

@ObjectType()
class QueryType {
	@Field()
	other(): QueryType {
		return null
	}
}

@Schema()
class SchemaType {
	@Query() query: QueryType
}

schemaFactory(SchemaType)
```

Results in:
```
RangeError: Maximum call stack size exceeded
```

### LINKED ISSUE #65: Circular Reference causes schema not to compile
Possibly related to Issue #64 

I have two types that refer to each other;
```
@ObjectType({ description: 'Store' })
class StoreType {
  @Field({ type: EmployeeType, isList: true })
  employees: EmployeeType[]
}

@ObjectType({ description: 'Employee' })
class EmployeeType {
  @Field({ type: StoreType })
  store: StoreType
}
```

As a result, I got this when trying to start the server;

```
{ Error: @ObjectType()'s store is annotated by @Field() but no type could be inferred
    at SchemaFactoryError (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/schema.type-factory.ts:27:5)
    at fieldMetadataList.reduce (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/object.type-factory.ts:52:21)
    at Array.reduce (native)
    at metadata_builder_1.getMetadataBuilder.buildObjectTypeMetadata.map.metadata (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/object.type-factory.ts:49:44)
    at Array.map (native)
    at Object.objectTypeFactory (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/object.type-factory.ts:22:8)
    at convertType (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/field.type-factory.ts:53:20)
    at Object.fieldTypeFactory (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/field.type-factory.ts:217:21)
    at fieldMetadataList.reduce (/Users/ryuuji3/Projects/shift-scheduler/node_modules/graphql-schema-decorator/src/type-factory/object.type-factory.ts:50:25)
    at Array.reduce (native) type: 2 }
```

When commenting out `Store.employees`, it compiles successfully. Then I uncommented `Store.employees` and commented out `Employee.store` and it compiled successfully.

Thank you!

### PR DESCRIPTION
**Pull request by VojtaStanek**
fix #64; fix #65 

It adds ability to specify type as a thunk as can be seen here: https://github.com/Cactucs/graphql-schema-decorator/blob/6ace1fb78883362ecfab87070a7a38c8b232daea/src/specs/object.type-factory.spec.ts#L84

### CODE REVIEW

File: src/type-factory/object.type-factory.ts [hunk 1]
```diff
@@ -46,7 +46,7 @@ export function objectTypeFactory(target: Function, isInput: boolean = false): g
                 SchemaFactoryErrorType.NO_FIELD);
           }
 
-          const fields = fieldMetadataList.reduce((map, fieldMetadata) => {
+          const fields = () => fieldMetadataList.reduce((map, fieldMetadata) => {
```
Comment (felipesabino) [inline src/type-factory/object.type-factory.ts:49]: 🤔 clever 👏 

I will leave this SO answer here for future reference on why this works https://stackoverflow.com/a/39290345/429521

**Review by felipesabino — APPROVED** [review]

**Comment by felipesabino** [comment]
@Cactucs 

Could you add an entry with you name and the change description to the [CHANGELOG.md](https://github.com/indigotech/graphql-schema-decorator/blob/master/CHANGELOG.md) file before I merge?

**Comment by Bennit** [comment]
Any update the status of this PR? I see there's another PR for the same issue opened 6 days ago.

**Comment by tibawatanabe** [comment]
@Cactucs @Bennit released 1.0.0-alpha10

**Pull request merged**